src/platform/webusb/enumeration.rs

```diff
@@ -0,0 +1,76 @@
+use wasm_bindgen_futures::{js_sys::Array, wasm_bindgen::JsCast, JsFuture};
+use web_sys::{UsbAlternateInterface, UsbConfiguration, UsbDevice, UsbInterface};
+
+use crate::{
+    descriptors::Configuration,
+    platform::webusb::device::{extract_decriptors, extract_string},
+    BusInfo, DeviceInfo, Error, InterfaceInfo,
+};
+
+pub async fn list_devices() -> Result<impl Iterator<Item = DeviceInfo>, Error> {
+    async fn inner() -> Result<Vec<DeviceInfo>, Error> {
+        let window = web_sys::window().unwrap();
+        let navigator = window.navigator();
+        let usb = navigator.usb();
+        let devices = JsFuture::from(usb.get_devices()).await.unwrap();
+
+        let devices: Array = JsCast::unchecked_from_js(devices);
+
+        let mut result = vec![];
+        for device in devices {
+            let device: UsbDevice = JsCast::unchecked_from_js(device);
+            JsFuture::from(device.open()).await.unwrap();
+
+            result.push(DeviceInfo {
+                bus_id: "webusb".to_string(),
+                device_address: 0,
+                vendor_id: device.vendor_id(),
+                product_id: device.product_id(),
+                device_version: ((device.usb_version_major() as u16) << 8)
+                    | device.usb_version_minor() as u16,
+                class: device.device_class(),
+                subclass: device.device_subclass(),
+                protocol: device.device_protocol(),
+                speed: None,
+                manufacturer_string: device.manufacturer_name(),
+                product_string: device.product_name(),
+                serial_number: device.serial_number(),
+                interfaces: {
```

Comment:
The problem is that this call returns empty alternates, hence the manual shenanigans with reading the descriptors.

Reply:
Hmm. I try to avoid device IO in `list_devices` because making requests to random devices can be slow, or even crash them. Less bad in WebUSB, though, because it's only devices selected at the browser prompt.

Reply:
Hmm I understand that concern. The question is how I get the respective webusb primitives to work then though :/

Reply:
I just tried it in Chrome on Linux from the JS console and get class/subclass/protocol but not the interface string descriptor. Is that what you're seeing, or is the interface entirely missing?

```
> d = await navigator.usb.requestDevice({filters: [{}]})
USBDevice {usbVersionMajor: 2, usbVersionMinor: 1, usbVersionSubminor: 0, deviceClass: 0, deviceSubclass: 0, …}
> d.configuration.interfaces[0].alternate
USBAlternateInterface {alternateSetting: 0, interfaceClass: 255, interfaceSubclass: 0, interfaceProtocol: 0, interfaceName: null, …}
```
That interface string descriptor is also missing on some driver installation configurations on Windows. Though you probably care about that string descriptor more than most because of CMSIS-DAP.

Reply:
Correct, that interface string descriptor is missing. For J-Link for example it's not relevant but the CMSIS-DAP spec requires us to check it.